Add Location Markers (Part 1)

AdenKoperczak opened the pull request:
A minimum viable implementation of location markers.
The location markers are on a new map layer. They can be configured from the location marker manager. They are saved to `location-markers.json` in the Appdata directory.

There are several more features that I would like to add, but I want to make sure I have a good starting point before I build this out more. Dan, feel free to review this as you have time.

AdenKoperczak committed:
initial poi manager and types code

AdenKoperczak committed:
Add poi layer

AdenKoperczak committed:
Add working poi manager implementation

AdenKoperczak committed:
Do initial work to get layer rendering

AdenKoperczak committed:
Merge branch 'dpaulat:develop' into points_of_interest

AdenKoperczak committed:
change texture for marker and remove unneeded logging

AdenKoperczak committed:
rename files to marker instead of poi

AdenKoperczak committed:
rename another file to marker fro poi

AdenKoperczak committed:
renamed all POI/point of intrest to marker

AdenKoperczak committed:
Reformat marker formats after rename

AdenKoperczak committed:
Updated how MarkerRecord works to make adding entryies easier

AdenKoperczak committed:
Added a basic location marker manager.

AdenKoperczak committed:
modified code to avoid cast from size_t to int properly

dpaulat commented on scwx-qt/source/scwx/qt/map/map_widget.cpp:
```diff
@@ -1232,6 +1235,12 @@ void MapWidgetImpl::AddLayer(types::LayerType        type,
                  { widget_->RadarSiteRequested(id); });
          break;
 
+      // Create the radar site layer
```
Copy/paste error, comment should be updated.

dpaulat commented on scwx-qt/source/scwx/qt/types/marker_types.hpp:
```diff
@@ -0,0 +1,26 @@
+
+struct MarkerInfo
+{
+   MarkerInfo(std::string name, double latitude, double longitude) :
```
Recommend `const std::string&` to avoid unnecessary copy

dpaulat commented on scwx-qt/source/scwx/qt/manager/marker_manager.cpp:
```diff
@@ -0,0 +1,282 @@
+      p->InitializeMarkerSettings();
+
+      // Read Marker settings on startup
+      // main::Application::WaitForInitialization();
```
The purpose of this is to make sure all slots/observers have been created and connected.  Otherwise, your MarkerModel might be out of sync on startup.  I recommend putting this block inside a `boost::asio::post` and uncommenting this line, similar to the PlacefileManager constructor, with associated thread pool logic.

AdenKoperczak replied:
I went back and forth on that one, but I see what you are saying now. I think I had MarkerModel load them on initialization, but your method is more robust if the ordering changes for some reason, so I will do that instead.

dpaulat commented on scwx-qt/source/scwx/qt/model/marker_model.cpp:
```diff
@@ -0,0 +1,255 @@
+      manager::MarkerManager::Instance()};
+};
+
+MarkerModel::MarkerModel(QObject* parent) :
```
Marker removal in the Marker Manager widget is sometimes delayed. The model should receive remove/update signals.
- On MarkerManager remove signal, emit `[begin|end]RemoveRows` from the model (see PlacefileModel)
- On MarkerManager update signal, emit `dataChanged` from the model (see PlacefileModel)
  - Recommend having MarkerManager emit which row is updated

dpaulat commented on scwx-qt/source/scwx/qt/manager/marker_manager.cpp:
```diff
@@ -0,0 +1,282 @@
+   Q_EMIT MarkersUpdated();
+}
+
+void MarkerManager::move_marker(size_t from, size_t to)
```
For a potential interface to this, see LayerModel. I implemented drag and drop using mime data (rather complicated), and then a few "simpler" arrow buttons on the side of the widget. This is probably OK for now, but I see this evolving some in Part 2+. I used `std::rotate` in LayerModel, similar to `remove_marker`, may have some performance optimizations.

AdenKoperczak replied:
This is not being used right now, and I think there will need to be an interface change anyway, so I will change it to use `std::rotate` when I get to moving the markers positions. Right now there is little need to move them, but I want to add hotkeys for "next/last marker" and that will make ordering more important.

dpaulat commented on scwx-qt/source/scwx/qt/manager/marker_manager.cpp:
```diff
@@ -0,0 +1,282 @@
+      return;
+   }
+
+   for (size_t i = index; i < p->markerRecords_.size() - 1; i++)
```
Recommend using `std::vector::erase` to erase specific element instead (i.e., `p->markerRecords_.erase(std::next(p->markerRecords_.begin(), i))`). May have some performance optimizations.

AdenKoperczak replied:
I misread how erase works. That is a lot simpler.

dpaulat commented on scwx-qt/source/scwx/qt/manager/marker_manager.cpp:
```diff
@@ -0,0 +1,282 @@
+std::shared_ptr<MarkerManager> MarkerManager::Instance()
+{
+   static std::weak_ptr<MarkerManager> markerManagerReference_ {};
+
```
Probably need an instance mutex (see `PlacefileManager`).

AdenKoperczak replied:
OK. To avoid the creation of multiple `*Manager` instances.

dpaulat commented on scwx-qt/source/scwx/qt/manager/marker_manager.cpp:
```diff
@@ -0,0 +1,282 @@
+   return p->markerRecords_.size();
+}
+
+// TODO deal with out of range/not found
```
I'll echo your TODO here. The MarkerModel relies on this returning some data.

Also, by returning a reference, is there a risk of a MarkerInfo reference being held onto after the MarkerRecord goes out of scope (shared_ptr reference count going to 0)? I typically don't prefer copies, but I'm wondering if `types::MarkerInfo` is a better return type here.

AdenKoperczak replied:
I think I will use `std::optional` here (without a reference), so I can indicate an out of range input without an exception.

dpaulat commented on scwx-qt/source/scwx/qt/manager/marker_manager.cpp:
```diff
@@ -0,0 +1,282 @@
+// TODO deal with out of range/not found
+const types::MarkerInfo& MarkerManager::get_marker(size_t index)
+{
+   std::shared_ptr<MarkerManager::Impl::MarkerRecord> markerRecord =
```
Recommend std::shared_ptr& to avoid unintended copy here.

dpaulat commented on scwx-qt/source/scwx/qt/manager/marker_manager.cpp:
```diff
@@ -0,0 +1,282 @@
+
+void MarkerManager::set_marker(size_t index, const types::MarkerInfo& marker)
+{
+   std::shared_ptr<MarkerManager::Impl::MarkerRecord> markerRecord =
```
Recommend std::shared_ptr& to avoid unintended copy here.

dpaulat commented on scwx-qt/source/scwx/qt/manager/marker_manager.cpp:
```diff
@@ -0,0 +1,282 @@
+   {
+      return;
+   }
+   std::shared_ptr<MarkerManager::Impl::MarkerRecord> markerRecord =
```
Recommend std::shared_ptr& to avoid unintended copy here.

dpaulat commented on scwx-qt/source/scwx/qt/ui/marker_settings_widget.cpp:
```diff
@@ -0,0 +1,105 @@
+static const std::string logPrefix_ = "scwx::qt::ui::marker_settings_widget";
+static const auto        logger_    = scwx::util::Logger::Create(logPrefix_);
+
+class MarkerSettingsWidgetImpl
```
Recommend having Name a bit wider, and Latitude/Longitude formatted similar to other dialogs.

AdenKoperczak replied:
Could you clarify what you are wanting form the Latitude/Longitude formatting? Are you looking for a spinbox format (this is a bit difficult), should the numbers be showing the trailing zeros when set to 0, or something else?

dpaulat replied:
I think the edit is just fine, but for the display role, I was thinking similar to how the radar site dialog will display using a formatted degrees N/S/E/W.

I should have an opportunity to review the rest this weekend.

AdenKoperczak replied:
Ok. I updated it to use the preexisting function for formatting. I agree that looks better.

dpaulat commented on scwx-qt/source/scwx/qt/map/marker_layer.cpp:
```diff
@@ -0,0 +1,108 @@
+
+   MarkerLayer* self_;
+   const std::string& markerIconName_ {
+      types::GetTextureName(types::ImageTexture::Cursor17)};
```
Eventually recommend a unique icon.  I've used Font Awesome Free and Flaticon.

AdenKoperczak replied:
That is defiantly just a temporary placeholder. I was planning on having a few selectable icons. Thanks for recommending a place to start looking.

dpaulat reviewed (changes requested):
Great work! Some of these comments may be able to be addressed in Part 2.

AdenKoperczak committed:
fix missed update of comment after copy/paste

AdenKoperczak committed:
use referance to avoid unnecessary copy

AdenKoperczak committed:
Add async code to marker_manager

AdenKoperczak committed:
Updated MarkerModel to update correctly on add/remove

AdenKoperczak committed:
Use std::vector::erase instead of self written code in remove_marker

AdenKoperczak committed:
Add proper bounds checks and update usage of references in marker_manager.cpp

AdenKoperczak committed:
Add modified methods to header file from last commit

AdenKoperczak committed:
Modify usage of get_marker for updated interface, and update checks for lat/lon input

AdenKoperczak committed:
update to propery check if string is empty

AdenKoperczak committed:
Updated marker to a custom marker so it is different from the center marker

AdenKoperczak commented:
I have fixed most of the comments. The only thing I still need to do is to get the UI looking a little better.
I switched to using an simple SVG I made for the markers for now. The traditional location pin implies the location is at the bottom center of the icon, so that would need to be taken into account using the offset, so I just skipped that for now.

AdenKoperczak committed:
add more handling to ensure that MarkerModel stays up to date with MarkerManager

AdenKoperczak committed:
Moved name to end of model to make it take all remaining space

AdenKoperczak committed:
fixed missing static cast on size_t to int convertion

AdenKoperczak committed:
Change cordinates to be displayed in a consistent format

dpaulat commented on scwx-qt/res/textures/images/location-marker.svg:
```diff
@@ -0,0 +1,9 @@
+<svg version="1.1"
+     width="15" height="15"
```
Looks slightly off of equilateral/isosceles. What if:
- 3s changed to 2s? I think it'd be slightly off center for what's in the middle of the triangle
- Width is increased to 16, and 12s can be changed to 13s?

dpaulat replied:
Although, 15x15 lends itself well to having a center pixel. I'm not sure the right answer here.

AdenKoperczak replied:
I have a fix using `viewBox` to make the coordinates.

dpaulat commented on scwx-qt/source/scwx/qt/manager/marker_manager.cpp:
```diff
@@ -191,19 +201,27 @@ size_t MarkerManager::marker_count()
 
-// TODO deal with out of range/not found
-const types::MarkerInfo& MarkerManager::get_marker(size_t index)
+std::optional<types::MarkerInfo> MarkerManager::get_marker(size_t index)
```
get/set/add/remove/move should take a mutex to avoid data races

AdenKoperczak replied:
I meant to ask if that was needed. Thanks for catching that.

dpaulat commented on scwx-qt/source/scwx/qt/map/marker_layer.cpp:
```diff
@@ -57,10 +57,14 @@ void MarkerLayer::Impl::ReloadMarkers()
+      geoIcons_->SetIconLocation(icon, marker->latitude, marker->longitude);
    }
 
    geoIcons_->FinishIcons();
```
Should emit the `NeedsRendering()` signal at the end of reloading markers. This will allow the layer to automatically update itself without having to wait for another event (e.g., mouseover or another trigger).

dpaulat commented on scwx-qt/source/scwx/qt/model/marker_model.cpp:
```diff
@@ -238,10 +252,22 @@ bool MarkerModel::setData(const QModelIndex& index,
+   beginInsertRows(QModelIndex(), 0, index);
+   endInsertRows();
+
+   Q_EMIT dataChanged(topLeft, bottomRight);
```
I don't believe you need `dataChanged` when you have `[begin|end]InsertRows`.

dpaulat commented on scwx-qt/source/scwx/qt/model/marker_model.cpp:
```diff
@@ -250,6 +276,27 @@ void MarkerModel::HandleMarkerAdded()
    Q_EMIT dataChanged(topLeft, bottomRight);
```
I don't believe you need `dataChanged` when you have `[begin|end]InsertRows`.

dpaulat commented on scwx-qt/source/scwx/qt/model/marker_model.cpp:
```diff
@@ -250,6 +276,27 @@ void MarkerModel::HandleMarkerAdded()
+   beginRemoveRows(QModelIndex(), removedIndex, removedIndex);
+   endRemoveRows();
+
+   Q_EMIT dataChanged(topLeft, bottomRight);
```
I don't believe you need `dataChanged` when you have `[begin|end]RemoveRows`.

dpaulat reviewed (commented):
One more round of comments. Thanks again!

AdenKoperczak committed:
Updated SVG to be much closer to an equilateral triangle

AdenKoperczak committed:
Add mutexes to marker manager

AdenKoperczak committed:
Make marker layer emit NeedsRendering on updating markers

AdenKoperczak committed:
removed unneded dataChanged calls in MarkerModel
